elf_reader.go

```diff
@@ -393,19 +395,51 @@ func (ec *elfCode) loadFunctions(section *elfSection) (map[string]asm.Instructio
 			// When splitting sections into subprograms, the targets of these calls
 			// are no longer in scope, so they must be resolved here.
 			if ins.IsFunctionReference() && ins.Constant != -1 {
-				tgt := jumpTarget(offset, ins)
+				tgt := jumpTarget(secOffset, ins)
 				sym := section.symbols[tgt].Name
 				if sym == "" {
-					return nil, fmt.Errorf("offset %d: no jump target found at offset %d", offset, tgt)
+					return nil, fmt.Errorf("offset %d: no jump target found at offset %d", secOffset, tgt)
 				}
 
-				ins.Reference = sym
+				ins.SetReference(sym)
 				ins.Constant = -1
 			}
 		}
 
+		// Only rebuild lineInfos at the start of a function
+		if funcOffset == 0 {
+			lineInfos = make(map[uint64]btf.LineInfo)
+			if ec.btf != nil {
+				for _, lineInfo := range ec.btf.GetFuncLineInfos(ins.Symbol()) {
+					lineInfos[uint64(lineInfo.InsnOff)] = lineInfo
+				}
+			}
+		}
+
+		if lineInfo, found := lineInfos[funcOffset]; found {
```

Comment:
Could a `String()` equivalent be implemented onto `btf.LineInfo` instead of having printing logic here? Might be easier to make rendering decisions, but maybe @lmb's PR (haven't reviewed yet) will make this a bit more straightforward.